zeppelin-web-angular/e2e/reporter.coverage.ts

```diff
@@ -0,0 +1,304 @@
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// @see https://playwright.dev/docs/test-reporters#custom-reporters
+import { FullResult, Reporter, TestCase, TestResult } from '@playwright/test/reporter';
+import { promises as fs } from 'fs';
+import { flatMap, sortBy } from 'lodash';
+import { scanDirectory, Results } from 'scandirectory';
+import cfg from './reporter.coverage.config';
+
+const TEST_STATUS = {
+  PASSED: 'passed',
+  SKIPPED: 'skipped',
+  FAILED: 'failed'
+} as const;
+
+type ResultsType = Array<[string, number, number, number, number, number]>;
+type TestStatusType = typeof TEST_STATUS[keyof typeof TEST_STATUS];
+interface TestedPathType {
+  success: number;
+  skipped: number;
+  failed: number;
+}
+
+const OUTPUT_FILE_NAME = 'coverage.log';
+const TABLE_COLUMNS = {
+  TOTAL: 1,
+  SUCCESS: 2,
+  FAILED: 3,
+  SKIPPED: 4
+} as const;
+
+class CoverageReporter implements Reporter {
+  testedPaths = new Map<string, TestedPathType>();
+  testedIds = new Map<string, TestStatusType>();
+  targetPaths: string[] = [];
+
+  async onBegin() {
+    console.log('Coverage reporter starting...');
+    console.log('Root path:', cfg.rootPath);
+
+    const results = await scanDirectory({
+      directory: cfg.rootPath
+    });
+
+    this.targetPaths = this.processScannedFiles(results);
+    console.log('Target paths:', this.targetPaths.length);
+  }
+
+  processScannedFiles(results: Results): string[] {
+    return Object.keys(results)
+      .filter(key => !results[key].directory)
+      .map(key => this.normalizeFilePath(key, results))
+      .filter(key => key !== '.')
+      .filter(key => this.shouldIncludeFile(key));
+  }
+
+  normalizeFilePath(key: string, results: Results): string {
+    if (/index\.tsx?$/.test(key)) {
+      return results[key].parent?.relativePath || '.';
+    }
+    return key.replace(/\.tsx?$/, '');
+  }
+
+  shouldIncludeFile(key: string): boolean {
+    if (cfg.testMatch?.length) {
+      const matchesTest = cfg.testMatch.some(rule => (rule instanceof RegExp ? rule.test(key) : rule === key));
+      if (!matchesTest) {
+        return false;
+      }
+    }
+
+    if (cfg.excludes?.length) {
+      const isExcluded = cfg.excludes.some(rule => (rule instanceof RegExp ? rule.test(key) : rule === key));
+      if (isExcluded) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  onTestEnd(test: TestCase, result: TestResult) {
+    const status =
+      result.status === TEST_STATUS.PASSED || result.status === TEST_STATUS.SKIPPED
+        ? (result.status as TestStatusType)
```

Comment:
I think the type assertion could be removed here, since the type is already narrowed by the ternary condition.